### ISSUE TITLE
Push Replication fails randomly

### ISSUE THREAD

**Issue by K-Leon** [couchbase/couchbase-lite-java-core#1614]
Hello,

I'm noticing something really strange: We've a small number of clients Running right now ( around 100 ) creating docs locally and pushing them to one remote sync gateway server and pull a few docs once in a while ( channel filtered replication ). Now i'm seeing from time to time clients that won't submit their newly created documents until the whole app restarts. It seems to happen especially if the client updated his Apk just before. ( App is Running - App Update - and now continue working with the app seems to fail often )

To make it short the Push Replication itself works but won't send any Data. 

While monitoring everything i've seen that replication requests got executed correctly but the client didn't tried to send anything until the bespoke app restart.

It's important to say that the client is parallel tapped in a pull replication feed which didn't have any updates for him.

Anyone has an Idea where to look for this issue? I'm seeing this nearly daily on a "random device". So it's really hard to tell where to look for.

Best Regards,
Leon

**Comment by hideki** [comment]
Hi @K-Leon,

Some questions:
1. Can you check Sync Gateway Log? Does Sync Gateway receive requests from clients?
2. Can you check LogCat on Android? Can you see any error message? and Can you see the indication of sending data to the sync gateway?

As you said, restarting the app can fix the problem. I wonder if the application loose internet connection or if replication is stopped after APK updated.

Please share further detailed information.

Thanks!

**Comment by K-Leon** [comment]
Hi @hideki 

I had luck and could start a Logcat tool on an instance that was already "broken". My findings: 

a) Internet Connection present - WIFI
b) Sync Gateway received requests - even though it had been POST request
c) I stopped Sync Gateway - looked at the logs and restarted it. I have seen connection failures coming from continious pull and of the One Shot Push Replication i ran additionally on a regular base to rule out issues with continious replication after seeing this issue the first time. Even though no PUT Request have been issued - only the local, Change Tracker requests etc. have been visible on a regular base. 
d) No Error message - nothing that indicates something is going wrong. At least nothing "continious" - i could only tap in when it did happen already. 

So this leaves me with two ideas:
a) Change Listener attached to Database failed
b) Mismatch between local and remote replication id.

As authentication i use basic auth.

Best Regards,
Leon

**Comment by K-Leon** [comment]
Hi @hideki,

I dove a little bit in the Code and i'm currently suspecting this part: https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java#L1020

Am I right: if there's an actual mismatch between both state the current revision isn't reset? I would expect that the replicator starts from the remote rev or starts over from the beginning ( rev 0 ).

Best Regards,
Leon

**Comment by hideki** [comment]
start over from `0`.

**Comment by K-Leon** [comment]
@hideki 
I just stumbled upon an instance with upgrade issues - this one was totally dead because upgrade caused an SQLiteDatabaseCorruptException.

Is it possible that the SQLite Database won't get closed correctly before updating?

Small snippet of the Log - it goes like this all the time: https://gist.github.com/K-Leon/28bda2a04970a09e29aaa76356cf8d8c

**Comment by hideki** [comment]
Hi @K-Leon,
Do you close the database in [`Application.onTerminate()`](https://developer.android.com/reference/android/app/Application.html#onTerminate()) method?

**Comment by K-Leon** [comment]
Apparently not: https://github.com/couchbaselabs/react-native-couchbase-lite/blob/master/android/src/main/java/me/fraserxu/rncouchbaselite/ReactCBLite.java

I'll look tomorrow into adding this - thank you very much! This could explain all of the weird behavior.

**Comment by K-Leon** [comment]
Sorry :) Now i started working on it and could read the docs - the whole Database won't get closed. Just added an onDestory Hook and i'll let you know if i'm still experiencing this behavior.

**Comment by K-Leon** [comment]
@hideki I dig a little bit in - but i haven't found a solution to close Databases on App Upgrade. 


03-24 11:12:46.768 790-812/? I/ActivityManager: Force stopping de.thisismy.app appid=10102 user=-1: uninstall pkg
03-24 11:12:46.769 790-812/? I/ActivityManager: Killing 10740:de.thisismy.app/u0a102 (adj 9): stop de.thisismy.app

This won't stop Couchbase Lite Manager - it keeps running and will easily fail while device moves Data around. Mostly nothing bad happens but from time to time we see strange things happen.

Is there a simple way to kill all already running instances before creating a new one or something like this?

**Comment by hideki** [comment]
@K-Leon,

1. Non of REST API methods explicitly close the database. We need to add the way to close the database through REST API.

https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/router/Router.java

2. 
Can you confirm [`Application.onTerminate()`](https://developer.android.com/reference/android/app/Application.html#onTerminate()) is called when Application is upgraded?
If yes, I think this is right time to make sure if database is closed.

**Comment by K-Leon** [comment]
Application.onTerminate ist never Called on Production Devices - so it's out. I tried all day to find a hook that allows execution of closing before running the upgrade without success.

The disturbing thing i noticed Lite Listener and Subsequently Couchbase Lite kept alive - even though App updated and didn't restarted Lite Listener / Couchbase Lite.

The best thing would be binding the CB / Lite Listener Lifetime to the Calling Process.

**Comment by hideki** [comment]
Hi @K-Leon,

I am not sure how application update behaves on Android. Also, I am not familiar with react-native. It seems Listener (web app for REST) threads become a deamon thread after the upgrade?

Do you use CBL 1.3.1 from react-native? Can you try 1.4.0 instead? We compiled SQLite serialized mode which allows multiple connections. It might prevent to cause Database Corruption.
https://github.com/couchbaselabs/react-native-couchbase-lite/blob/master/android/build.gradle

**Comment by K-Leon** [comment]
Hi @hideki

I'm already on 1.4

**Comment by K-Leon** [comment]
Yes Listener somehow Deamonizes - RN isn't really different to all other apps using rest API.

**Comment by hideki** [comment]
Hi @jamiltz,
do you have any idea about this issue?
